eis_toolkit/transformations/coda/ilr.py

```diff
@@ -0,0 +1,104 @@
+
+    Returns:
+        The scaling factor.
+
+    Raises:
+        InvalidParameterValueException: One or both of the input values are zero or negative.
+    """
+    if not (check_numeric_value_sign(c1) and check_numeric_value_sign(c2)):
+        raise InvalidParameterValueException("Input values must both be positive integers.")
+
+    return np.sqrt((c1 * c2) / np.float64(c1 + c2))
+
+
+@beartype
+def _geometric_mean_logratio(
+    row: pd.Series, subcomposition_1: Sequence[str], subcomposition_2: Sequence[str]
+) -> np.float64:
+
+    numerator = gmean(row[subcomposition_1])
+    denominator = gmean(row[subcomposition_2])
+    return np.log(numerator / denominator)
+
+
+@beartype
+def _single_ilr_transform(
+    df: pd.DataFrame, subcomposition_1: Sequence[str], subcomposition_2: Sequence[str]
+) -> pd.Series:
+
+    dfc = df.copy()
+
+    c1 = len(subcomposition_1)
+    c2 = len(subcomposition_2)
+
+    # A Series to hold the transformed rows
+    ilr_values = pd.Series([0.0] * df.shape[0])
+
+    for idx, row in dfc.iterrows():
+        ilr_values[idx] = _geometric_mean_logratio(row, subcomposition_1, subcomposition_2)
+
+    ilr_values = _calculate_ilr_scaling_factor(c1, c2) * ilr_values
+
+    return ilr_values
+
+
+@beartype
+@check_compositional
+def single_ilr_transform(
+    df: pd.DataFrame, subcomposition_1: Sequence[str], subcomposition_2: Sequence[str]
+) -> pd.Series:
+    """
+    Perform a single isometric logratio transformation on the provided subcompositions.
+
+    Returns ILR balances. Column order matters.
+
+    Args:
+        df: A dataframe of shape [N, D] of compositional data.
+        subcomposition_1: Names of the columns in the numerator part of the ratio.
+        subcomposition_2: Names of the columns in the denominator part of the ratio.
+
+    Returns:
+        A series of length N containing the transforms.
+
+    Raises:
+        InvalidColumnException: One or more subcomposition columns are not found in the input dataframe.
+        InvalidCompositionException: One or more columns are found in both subcompositions.
+        InvalidParameterValueException: At least one subcomposition provided was empty.
+        See check_compositional for other exceptions.
+    """
+    if not (subcomposition_1 and subcomposition_2):
+        raise InvalidParameterValueException("A subcomposition should contain at least one column.")
+
+    if not (check_columns_valid(df, subcomposition_1) and check_columns_valid(df, subcomposition_2)):
+        raise InvalidColumnException("Not all of the input columns were found in the input dataframe.")
+
+    if check_lists_overlap(subcomposition_1, subcomposition_2):
+        raise InvalidCompositionException("The subcompositions overlap.")
+
+    return _single_ilr_transform(df, subcomposition_1, subcomposition_2)
+
+
+@beartype
+def inverse_ilr():
+    """Perform the inverse transformation for a set of ILR transformed data."""
+    raise NotImplementedError()
```

Comment:
Will this be implemented? If not (if it is not a defined operation or something like that), I would leave it out.

Reply:
This should be implemented according some excel sheet I found on Teams, so I left it as a memo. But as we still don't even know how the full set of ILR transforms should be produced, might as well remove this as well. I'll clean it up 👍